tarantool/msgpack_ext/types/datetime.py

```diff
@@ -0,0 +1,247 @@
+#   little-endian order.
+# * The optional fields (8 bytes), if any of them have a non-zero value.
+#   The fields include nsec (4 bytes), tzoffset (2 bytes), and
+#   tzindex (2 bytes) packed in the little-endian order.
+#
+# seconds is seconds since Epoch, where the epoch is the point where the time
+# starts, and is platform dependent. For Unix, the epoch is January 1,
+# 1970, 00:00:00 (UTC). Tarantool uses a double type, see a structure
+# definition in src/lib/core/datetime.h and reasons in
+# https://github.com/tarantool/tarantool/wiki/Datetime-internals#intervals-in-c
+#
+# nsec is nanoseconds, fractional part of seconds. Tarantool uses int32_t, see
+# a definition in src/lib/core/datetime.h.
+#
+# tzoffset is timezone offset in minutes from UTC. Tarantool uses a int16_t type,
+# see a structure definition in src/lib/core/datetime.h.
+#
+# tzindex is Olson timezone id. Tarantool uses a int16_t type, see a structure
+# definition in src/lib/core/datetime.h. If both tzoffset and tzindex are
+# specified, tzindex has the preference and the tzoffset value is ignored.
+
+SECONDS_SIZE_BYTES  = 8
+NSEC_SIZE_BYTES     = 4
+TZOFFSET_SIZE_BYTES = 2
+TZINDEX_SIZE_BYTES  = 2
+
+BYTEORDER = 'little'
+
+NSEC_IN_SEC = 1000000000
+SEC_IN_MIN = 60
+MIN_IN_DAY = 60 * 24
+
+
+def get_bytes_as_int(data, cursor, size):
+    part = data[cursor:cursor + size]
+    return int.from_bytes(part, BYTEORDER, signed=True), cursor + size
+
+def get_int_as_bytes(data, size):
+    return data.to_bytes(size, byteorder=BYTEORDER, signed=True)
+
+def compute_offset(timestamp):
+    if timestamp.tz is None:
+        return 0
+
+    utc_offset = timestamp.tz.utcoffset(timestamp)
+    # There is no precision loss since offset is in minutes
+    return utc_offset.days * MIN_IN_DAY + utc_offset.seconds // SEC_IN_MIN
+
+def pytz_to_tarantool(timestamp):
+# Tarantool 2.10 (commit 9ee45289e01232b8df1413efea11db170ae3b3b4)
+# do not support the following pytz (version 2022.2.1) timezones
+# - CST6CDT
+# - EST5EDT
+# - Etc/GMT+1
+# - Etc/GMT+10
+# - Etc/GMT+11
+# - Etc/GMT+12
+# - Etc/GMT+2
+# - Etc/GMT+3
+# - Etc/GMT+4
+# - Etc/GMT+5
+# - Etc/GMT+6
+# - Etc/GMT+7
+# - Etc/GMT+8
+# - Etc/GMT+9
+# - Etc/GMT-1
+# - Etc/GMT-10
+# - Etc/GMT-11
+# - Etc/GMT-12
+# - Etc/GMT-13
+# - Etc/GMT-14
+# - Etc/GMT-2
+# - Etc/GMT-3
+# - Etc/GMT-4
+# - Etc/GMT-5
+# - Etc/GMT-6
+# - Etc/GMT-7
+# - Etc/GMT-8
+# - Etc/GMT-9
+# - Europe/Kyiv
+# - MET
+# - MST7MDT
+# - PST8PDT
+#
+# They are transformed to tzoffset based on pytz info.
+    tzoffset = compute_offset(timestamp)
+
+    if (timestamp.tz is None) or (timestamp.tz.zone is None):
+        tzindex = 0
+    else:
+        if timestamp.tz.zone in tt_timezones.timezoneToIndex:   
+            tzindex = tt_timezones.timezoneToIndex[timestamp.tz.zone]   
+        else:  
+            warn(f'pytz timezone {timestamp.tz} is not supported by Tarantool, '
+                 f'using tzoffset={tzoffset} instead', MsgpackWarning)
```

Comment:
It seems to me that an error is more appropriate here: we are not encode what a user expects.

Reply:
Reworked to raise erros